chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2222 @@
+        raw_tx = ""
+        for i in range(0, 4):
+            raw_tx += _mem.txfreq[i].get_raw()
+        return raw_tx == "\xFF\xFF\xFF\xFF"
+
+    # Encoding processing
+    def _decode_tone(self, val):
+        if val == 16665:
+            return '', None, None
+        elif val >= 12000:
+            return 'DTCS', val - 12000, 'R'
+        elif val >= 8000:
+            return 'DTCS', val - 8000, 'N'
+        else:
+            return 'Tone', val / 10.0, None
+
+    # Decoding processing
+    def _encode_tone(self, memval, mode, value, pol):
+        if mode == "":
+            memval[0].set_raw(0xFF)
+            memval[1].set_raw(0xFF)
+        elif mode == 'Tone':
+            memval.set_value(int(value * 10))
+
+        elif mode == 'DTCS':
+            flag = 0x80 if pol == 'N' else 0xC0
+            memval.set_value(value)
+            memval[1].set_bits(flag)
+        else:
+            raise Exception("Internal error: invalid mode `%s'" % mode)
+
+    def _get_mem(self, number):
+        return self._memobj.memory[number]
+
+    def _get_nam(self, number):
+        return self._memobj.names[number]
+
+    def _get_fm(self, number):
+        return self._memobj.fmmode[number]
+
+    def _get_get_scanvfo(self, number):
+        return self._memobj.fmvfo[number]
+
+    def _get_scan(self, number):
+        return self._memobj.scanadd[number]
+
+    def _get_block_data(self, number):
+        return self._memobj.blockfirm[number]
+
+    def _get_fmblock_data(self, number):
+        return self._memobj.fmblockfirm[number]
+
+    def get_memory(self, number):
+        _mem = self._get_mem(number)
+        _nam = self._get_nam(number)
+        mem = chirp_common.Memory()
+
+        mem.number = number
+
+        if _mem.get_raw()[0] == "\xff":
+            mem.empty = True
+            return mem
+
+        # Receiving frequency
+        mem.freq = int(_mem.rxfreq) * 10
+
+        # Offset
+        if self._gmrs:
+            mem.duplex = ""
+            mem.offset = 0
+        else:
+            if self._is_txinh(_mem):
+                mem.duplex = ""
+                mem.offset = 0
+            elif int(_mem.rxfreq) == int(_mem.txfreq):
+                mem.duplex = ""
+                mem.offset = 0
+            elif abs(int(_mem.rxfreq) * 10 - int(_mem.txfreq) * 10) > 70000000:
+                mem.duplex = "split"
+                mem.offset = int(_mem.txfreq) * 10
+            else:
+                mem.duplex = int(_mem.rxfreq) > int(_mem.txfreq) and "-" or "+"
+                mem.offset = abs(int(_mem.rxfreq) - int(_mem.txfreq)) * 10
+
+        # Channel name
+        for char in _nam.name:
+            if str(char) == "\xFF":
+                char = " "
+            mem.name += str(char)
+        mem.name = mem.name.rstrip()
+        if (self.radio_ident == "P31185" or self.radio_ident == "P31184") and \
+                (mem.number >= 189 and mem.number <= 199):
+            mem.name = HAM_NAME[mem.number - 200]
```

Comment:
It would really be better to delegate this sort of thing to a handler in the subclasses. Something like:
```
mem.name = self._get_model_name(mem.number, _nam)

def _get_model_name(self, number, _nam):
    return str(_nam.name).replace('\xFF', '').rstrip()
```
and then define this in the ham-specific class:
```
def _get_model_name(self, number, _nam):
    if 189 <= number <= 199:
        mem.name = HAM_NAME[number - 200]
    else:
        mem.name = super()._get_model_name(number, _nam)
```

Reply:
Thank you for the method you provided, because he uses this HAM_NAME in both HAM and GMRS, so I should modify the name of this HAM_NAME to avoid misleading, so I should write it like this:

```
mem.name = mem.name.rstrip()
if self.ident_mode != b'P31183\xff\xff' and \
        (mem.number >= 189 and mem.number <= 199):
    mem.name = HAM_NAME[mem.number - 200]
```